### PR TITLE
[17598] Incorporating parallel rspec gem to allow for parallel rspec …

### PR DESCRIPTION
**Pull request by bosawt**
…tests locally

## Description of change
This PR adds the `parallel-rspec` Gem, and the proper settings update to incorporate the ability to parallelize Rspec tests locally. This PR also fixes multiple spec failures that occur when the order of files that are run as spec tests are changed.

## Original issue(s)
department-of-veterans-affairs/va.gov-team/issues/17598

## Things to know about this PR
- This PR changes the name of the test database, it's currently `vets-api-test` in `settings.yml`, but `vets_api_test` in the Docker setup, so this makes it consistent, `vets_api_test` everywhere.
- Because of the test database name change, this could potentially temporarily affect everyone, they'll have to run a `RAILS_ENV=test bundle exec rake db:reset` to fix it. Make sure to `RAILS_ENV=test` so that your own development database is not dropped.
- Besides the test database name change, running `bundle exec rspec` or `make spec` should operate exactly as it did before.
- I removed the `?pool=4` option for the test database in the docker compose. Is this even necessary? It looks like we set the pool connections to 5 in the `database.yml` as a default anyway. Perhaps we can change the pool connections in that file, if necessary?

### Local Setup
- To get this running locally:
- `RAILS_ENV=test bundle exec rake db:reset`
- `bundle exec rake parallel:setup`
THEN, when testing with RSpec:
- `bundle exec rake parallel:spec`

### Docker Setup
- This PR adds two commands to the docker setup, `spec_parallel_setup`, and `spec_parallel`
- `make spec_parallel_setup` -> To reset the base test database, and build the parallel databases
- `make spec_parallel SPEC_PATH=path/to/spec/file/or/directory` -> Run the parallel specs!

### CODE REVIEW

**Comment by AGarcia-Clarity** [comment]
The code LGTM and I tested it out with the Docker instructions.
Recommendations:
- Update the documentation for both the [Native (local)](https://github.com/department-of-veterans-affairs/vets-api/blob/master/docs/setup/running_natively.md) commands & [Docker commands](https://github.com/department-of-veterans-affairs/vets-api/blob/master/docs/setup/running_docker.md) commands to reflect the changes.

File: spec/lib/bip_claims/service_spec.rb [hunk 1]
```diff
@@ -2,6 +2,7 @@
 
 require 'rails_helper'
 require 'bip_claims/service'
+require 'mpi/attr_service'
```
Comment (erluti) [inline spec/lib/bip_claims/service_spec.rb:5]: This should actually go in `lib/bip_claims/service.rb`

**Comment by annaswims** [comment]
Even if you don't merge parallel tests right away it'd be nice to have the missing requires added soon.  I just came across one when I was trying to run just `spec/request/swagger_spec.rb`

File: spec/lib/bip_claims/service_spec.rb (hunk 1 above)
Comment (bosawt) [reply spec/lib/bip_claims/service_spec.rb]: 👍

**Comment by bosawt** [comment]
> Even if you don't merge parallel tests right away it'd be nice to have the missing requires added soon. I just came across one when I was trying to run just `spec/request/swagger_spec.rb`

Sounds good, if there aren't issues with the 

> The code LGTM and I tested it out with the Docker instructions.
> Recommendations:
> 
> * Update the documentation for both the [Native (local)](https://github.com/department-of-veterans-affairs/vets-api/blob/master/docs/setup/running_natively.md) commands & [Docker commands](https://github.com/department-of-veterans-affairs/vets-api/blob/master/docs/setup/running_docker.md) commands to reflect the changes.

Cool, I updated the documentation to reflect the new commands

**Comment by bramleyjl** [comment]
LGTM, running on native installation

**Review by bramleyjl — APPROVED** [review]

**Comment by bosawt** [comment]
> Even if you don't merge parallel tests right away it'd be nice to have the missing requires added soon. I just came across one when I was trying to run just `spec/request/swagger_spec.rb`

It seems like this is working, so I'm moving it to a 'real' PR. If there's pushback on the parallel spec component I'll split out the spec requires and get that in. Otherwise if this continues moving forward I'll keep it together for this PR.

File: config/settings.yml [hunk 2]
```diff
@@ -57,7 +57,7 @@ binaries:
 db_encryption_key: f01ff8ebd1a2b053ad697ae1f0d86adb
 
 database_url: postgis:///vets-api
-test_database_url: postgis:///vets-api-test
+test_database_url: postgis://postgres:password@postgres:5432/vets_api_test
```
Comment (annaswims) [inline config/settings.yml:60]: is this change absolutely necessary?  I'll require everyone to create this db user with this publically available password and many/most people  don't use docker.

Comment (bosawt) [reply config/settings.yml:60]: So it needs to match the `test_database_url` pattern in the `docker_compose.yml` file. But, I had a thought. What If I left this line unchanged, but in the `README` told potential users to add `test_database_url: postgis://postgres:password@postgres:5432/vets_api_test` in `settings.local.yml`. A bit tacky, but not end of the world.

Comment (bosawt) [reply config/settings.yml]: Actually, @annaswims, mind reviewing the latest commit? It turns out, when I use the command `parallel_tests` it lets me run a command in parallel while persisting the environment settings (in this case, specifically the Settings override from the docker-compose file). So I'm able to just create the databases with that command, and it appears to work. Doing this, I don't need to change `Settings.yml` at all!

**Comment by erluti** [comment]
When trying to run the docker `make spec_parallel_setup` command, I see this error a bunch `Errno::EPERM: Operation not permitted - bs_fetch:atomic_write_cache_file:chmod` but it looks like it finished ok.  But running `make spec_parallel` is showing the same error as well as errors for `vets_api_test2` and `vets_api_test4` not existing.

**Comment by bosawt** [comment]
> When trying to run the docker `make spec_parallel_setup` command, I see this error a bunch `Errno::EPERM: Operation not permitted - bs_fetch:atomic_write_cache_file:chmod` but it looks like it finished ok. But running `make spec_parallel` is showing the same error as well as errors for `vets_api_test2` and `vets_api_test4` not existing.

Ah yeah, I originally saw that, but it seemed to resolve itself and I wasn't able to reproduce it. That's an error from Bootsnap, although it could either be an actual Bootsnap issue, or due to an issue with the creation of the databases. Do you mind commenting out `require 'bootsnap/setup'` in `boot.rb` and trying again?

**Comment by ericboehs** [comment]
One thing I found odd is that this requires the vets-api database to exist. `rails db:create` fixes the error but I'm worried it's modifying something in there.

Another thing, I kept getting a message saying rails db:migrate needed to be ran. I couldn't get it to go away but Phillip helped me find the correct commands:

```
bundle exec rails db:reset
RAILS_ENV=test bundle exec rake db:reset
bundle exec rake parallel:drop
bundle exec rake parallel:setup
bundle exec rake parallel:spec
```

**Comment by bosawt** [comment]
> One thing I found odd is that this requires the vets-api database to exist. `rails db:create` fixes the error but I'm worried it's modifying something in there.
> 
> Another thing, I kept getting a message saying rails db:migrate needed to be ran. I couldn't get it to go away but Phillip helped me find the correct commands:
> 
> ```
> bundle exec rails db:reset
> RAILS_ENV=test bundle exec rake db:reset
> bundle exec rake parallel:drop
> bundle exec rake parallel:setup
> bundle exec rake parallel:spec
> ```

Ah yeah, the way it works is that it creates a copy of your dev database for the different test databases. `rake parallel:drop/setup` will never actually touch the development database. But it's a good call to mention that it has to exist. I'll update the docs.

**Comment by erluti** [comment]
> > When trying to run the docker `make spec_parallel_setup` command, I see this error a bunch `Errno::EPERM: Operation not permitted - bs_fetch:atomic_write_cache_file:chmod` but it looks like it finished ok. But running `make spec_parallel` is showing the same error as well as errors for `vets_api_test2` and `vets_api_test4` not existing.
> 
> Ah yeah, I originally saw that, but it seemed to resolve itself and I wasn't able to reproduce it. That's an error from Bootsnap, although it could either be an actual Bootsnap issue, or due to an issue with the creation of the databases. Do you mind commenting out `require 'bootsnap/setup'` in `boot.rb` and trying again?

Commenting that out and it's running without the error

**Comment by bosawt** [comment]
> > > When trying to run the docker `make spec_parallel_setup` command, I see this error a bunch `Errno::EPERM: Operation not permitted - bs_fetch:atomic_write_cache_file:chmod` but it looks like it finished ok. But running `make spec_parallel` is showing the same error as well as errors for `vets_api_test2` and `vets_api_test4` not existing.
> > 
> > 
> > Ah yeah, I originally saw that, but it seemed to resolve itself and I wasn't able to reproduce it. That's an error from Bootsnap, although it could either be an actual Bootsnap issue, or due to an issue with the creation of the databases. Do you mind commenting out `require 'bootsnap/setup'` in `boot.rb` and trying again?
> 
> Commenting that out and it's running without the error

Dang, good news for this PR, but means something ain't kosher with bootsnap. If the `tmp/cache` directory is deleted, bootsnap will have to recreate the files from scratch, and I expected the problem may be resolved, but that's obviously not a great solution for people in general

File: docs/setup/running_natively.md [hunk 3]
```diff
@@ -19,6 +19,12 @@ they would be when running rails directly.
 - `bundle exec rake spec` - Run the entire test suite  ( for `rspec spec`). Test coverage statistics are in `coverage/index.html` or in [CodeClimate](https://codeclimate.com/github/department-of-veterans-affairs/vets-api/code)
 - `make guard` - Run the guard test server that reruns your tests after files are saved. Useful for TDD!
 
+### Running tests in parallel
+- NOTE: Running specs in parallel requires that your development database exists, and is up to date. If necessary, you may need to run `bundle exec rake db:create` and `bundle exec rake db:migrate` before the following steps
+- `RAILS_ENV=test bundle exec rake db:reset` - Set up your test database properly first. This just makes sure if your existing test database is named differently than what is expected, it will be resolved first
```
Comment (annaswims) [inline docs/setup/running_natively.md:24]: I don't think `rake db:reset` is necessary now that you've reverted the change in settings.yml

**Comment by annaswims** [comment]
When I ran locally I got a couple errors
```
 1) Form1010cg::DeliverPdfToCARMAJob#perform unit cleanup when PDF is still present and PDF deletion fails behaves like a successful job processes attachment
     Got 1 failure and 1 other error:
     Shared Example Group: "a successful job" called from ./spec/workers/form1010cg/deliver_pdf_to_carma_job_spec.rb:134

     1.1) Failure/Error: require_relative 'base'

          NameError:
            uninitialized constant RSpec::Support::Differ
          # ./lib/carma/models/submission.rb:4:in `<top (required)>'
          # ./app/services/form1010cg/service.rb:5:in `<top (required)>'
          # ./spec/workers/form1010cg/deliver_pdf_to_carma_job_spec.rb:33:in `block (5 levels) in <top (required)>'
          # ------------------
          # --- Caused by: ---
          # NameError:
          #   uninitialized constant RSpec::Support::Differ
          #   ./lib/carma/models/submission.rb:4:in `<top (required)>'

     1.2) Failure/Error: expect { submission.reload }.to raise_error(ActiveRecord::RecordNotFound)
            Expected block to raise error #<ActiveRecord::RecordNotFound>.
          # ./spec/workers/form1010cg/deliver_pdf_to_carma_job_spec.rb:52:in `block (5 levels) in <top (required)>'

  2) Form1010cg::DeliverPdfToCARMAJob#perform unit cleanup when PDF is still present and PDF deletion succeeds behaves like a successful job processes attachment
     Got 1 failure and 1 other error:
     Shared Example Group: "a successful job" called from ./spec/workers/form1010cg/deliver_pdf_to_carma_job_spec.rb:121

     2.1) Failure/Error: require_relative 'base'

          NameError:
            uninitialized constant RSpec::Support::Differ
          # ./lib/carma/models/submission.rb:4:in `<top (required)>'
          # ./app/services/form1010cg/service.rb:5:in `<top (required)>'
          # ./spec/workers/form1010cg/deliver_pdf_to_carma_job_spec.rb:33:in `block (5 levels) in <top (required)>'
          # ------------------
          # --- Caused by: ---
          # NameError:
          #   uninitialized constant RSpec::Support::Differ
          #   ./lib/carma/models/submission.rb:4:in `<top (required)>'

     2.2) Failure/Error: expect { submission.reload }.to raise_error(ActiveRecord::RecordNotFound)
            Expected block to raise error #<ActiveRecord::RecordNotFound>.
          # ./spec/workers/form1010cg/deliver_pdf_to_carma_job_spec.rb:52:in `block (5 levels) in <top (required)>'

Finished in 2 minutes 16.3 seconds (files took 21.82 seconds to load)
627 examples, 2 failures, 3 pending
```

Probably need to run this several more times just to make sure there aren't more that intermittently fail.

**Comment by annaswims** [comment]
I don't think it matters too much since it seems to work on Jenkins, but locally simplecov isn't reporting the correct coverage. 
Though maybe test run 6 was the one with the errors and didn't report coverage. 
```
Coverage report generated for (1/8), (2/8), (3/8), (4/8), (5/8), (7/8), (8/8) to /Users/annacarey/code/vets-api/coverage. 28914 / 47000 LOC (61.52%) covered.
File (/Users/annacarey/code/vets-api/modules/vaos/app/models/vaos/cancel_form.rb) is only (0.00%) covered. This is below the expected minimum coverage per file of (90.00%).
SimpleCov failed with exit 2
5920 examples, 2 failures, 9 pendings

Took 290 seconds (4:50)
Tests Failed
```

File: docs/setup/running_natively.md (hunk 3 above)
Comment (bosawt) [reply docs/setup/running_natively.md:24]: Sounds good, I'll make this change to remove the line

**Comment by bosawt** [comment]
> I don't think it matters too much since it seems to work on Jenkins, but locally simplecov isn't reporting the correct coverage.
> Though maybe test run 6 was the one with the errors and didn't report coverage.
> 
> ```
> Coverage report generated for (1/8), (2/8), (3/8), (4/8), (5/8), (7/8), (8/8) to /Users/annacarey/code/vets-api/coverage. 28914 / 47000 LOC (61.52%) covered.
> File (/Users/annacarey/code/vets-api/modules/vaos/app/models/vaos/cancel_form.rb) is only (0.00%) covered. This is below the expected minimum coverage per file of (90.00%).
> SimpleCov failed with exit 2
> 5920 examples, 2 failures, 9 pendings
> 
> Took 290 seconds (4:50)
> Tests Failed
> ```
> 
> Edit: a failed tests is not the reason
> 
> ```
> Coverage report generated for (1/8), (2/8), (3/8), (4/8), (5/8), (6/8), (7/8), (8/8) to /Users/annacarey/code/vets-api/coverage. 29425 / 46828 LOC (62.84%) covered.
> File (/Users/annacarey/code/vets-api/modules/vaos/app/serializers/vaos/v0/availability_serializer.rb) is only (0.00%) covered. This is below the expected minimum coverage per file of (90.00%).
> SimpleCov failed with exit 2
> 5920 examples, 0 failures, 9 pendings|
> ```

I ran a standard full RSpec run, and it looks like I get the same thing:
```
Coverage report generated for RSpec to /Users/bosaw/Work/Oddball/vets-api/coverage. 29425 / 46828 LOC (62.84%) covered.
File (/Users/bosaw/Work/Oddball/vets-api/modules/vaos/app/serializers/vaos/v0/availability_serializer.rb) is only (0.00%) covered. This is below the expected minimum coverage per file of (90.00%).
SimpleCov failed with exit 2%
```

What do you think I should do here? I can ignore it, since this is the existing behavior for a full run, or I can update the docs to recommend adding the `NOCOVERAGE` flag when doing a parallel run

**Comment by bosawt** [comment]
> When I ran locally I got a couple errors
> 
> ```
>  1) Form1010cg::DeliverPdfToCARMAJob#perform unit cleanup when PDF is still present and PDF deletion fails behaves like a successful job processes attachment
>      Got 1 failure and 1 other error:
>      Shared Example Group: "a successful job" called from ./spec/workers/form1010cg/deliver_pdf_to_carma_job_spec.rb:134
> 
>      1.1) Failure/Error: require_relative 'base'
> 
>           NameError:
>             uninitialized constant RSpec::Support::Differ
>           # ./lib/carma/models/submission.rb:4:in `<top (required)>'
>           # ./app/services/form1010cg/service.rb:5:in `<top (required)>'
>           # ./spec/workers/form1010cg/deliver_pdf_to_carma_job_spec.rb:33:in `block (5 levels) in <top (required)>'
>           # ------------------
>           # --- Caused by: ---
>           # NameError:
>           #   uninitialized constant RSpec::Support::Differ
>           #   ./lib/carma/models/submission.rb:4:in `<top (required)>'
> 
>      1.2) Failure/Error: expect { submission.reload }.to raise_error(ActiveRecord::RecordNotFound)
>             Expected block to raise error #<ActiveRecord::RecordNotFound>.
>           # ./spec/workers/form1010cg/deliver_pdf_to_carma_job_spec.rb:52:in `block (5 levels) in <top (required)>'
> 
>   2) Form1010cg::DeliverPdfToCARMAJob#perform unit cleanup when PDF is still present and PDF deletion succeeds behaves like a successful job processes attachment
>      Got 1 failure and 1 other error:
>      Shared Example Group: "a successful job" called from ./spec/workers/form1010cg/deliver_pdf_to_carma_job_spec.rb:121
> 
>      2.1) Failure/Error: require_relative 'base'
> 
>           NameError:
>             uninitialized constant RSpec::Support::Differ
>           # ./lib/carma/models/submission.rb:4:in `<top (required)>'
>           # ./app/services/form1010cg/service.rb:5:in `<top (required)>'
>           # ./spec/workers/form1010cg/deliver_pdf_to_carma_job_spec.rb:33:in `block (5 levels) in <top (required)>'
>           # ------------------
>           # --- Caused by: ---
>           # NameError:
>           #   uninitialized constant RSpec::Support::Differ
>           #   ./lib/carma/models/submission.rb:4:in `<top (required)>'
> 
>      2.2) Failure/Error: expect { submission.reload }.to raise_error(ActiveRecord::RecordNotFound)
>             Expected block to raise error #<ActiveRecord::RecordNotFound>.
>           # ./spec/workers/form1010cg/deliver_pdf_to_carma_job_spec.rb:52:in `block (5 levels) in <top (required)>'
> 
> Finished in 2 minutes 16.3 seconds (files took 21.82 seconds to load)
> 627 examples, 2 failures, 3 pending
> ```
> 
> Probably need to run this several more times just to make sure there aren't more that intermittently fail.

Also, I did try this several times with different parallel threads (so different file lists). I wasn't able to reproduce this, were you able to fix it on your local machine? I can include the change in this PR if so

**Review by ericboehs — APPROVED** [review]
This is working for me.

**Comment by bosawt** [comment]
@annaswims @erluti Any objections to this being merged? I believe I've addressed the `Bootsnap` related feedback from @erluti. And with @annaswims 's feedback, the SimpleCov report failure appears to be an existing problem (not related to parallel specs), and the flaky spec was one that I wan't able to reproduce (my test process was running the entire suite once each with max number of processors set from 2-16).
In any case, nothing in this PR should impact CI, or anyone running specs in the way they do today. So if there continue to be a flaky spec with a parallel run, we should be able to fix it without impacting anyone's existing workflow.

Also, wanted to mention a thank you for all the iteration on this PR. Sorry it probably took so much of all of your time, but I think the changes from the feedback have made this much better already.

**Comment by ericboehs** [comment]
I didn't see a bug in simplecov. Coverage results you posted looked correct as well as testing locally.

**Comment by erluti** [comment]
> @annaswims @erluti Any objections to this being merged?
It's looking good to me.  I'd like the conflicts to be resolved and it sorted out what state simple cov is in before I officially 👍 .
I think it's okay for this PR that simplecov doesn't run for the parallel tests (target is local develpment, right?), but everything should be clear to the user

**Comment by bosawt** [comment]
> @annaswims @erluti Any objections to this being merged?
> It's looking good to me. I'd like the conflicts to be resolved and it sorted out what state simple cov is in before I officially 👍 .
> I think it's okay for this PR that simplecov doesn't run for the parallel tests (target is local develpment, right?), but everything should be clear to the user

Sounds good, I added documentation to suggest adding the `NOCOVERAGE=true` flag when running parallel specs. That flag is already there for the Docker parallel spec command as well.

**Review by erluti — APPROVED** [review]
Thanks for all this work!

**Pull request merged**